### PR TITLE
Propagate to Lambda type information for function parameters and return

### PR DESCRIPTION
**Pull request by chambart**
Type information about function parameter and return can be useful for various optimizations on flambda and clambda. In particular it helps function parameter unboxing and post-inlining primitive specialization.

@alainfrisch you might be concerned by that.

### CODE REVIEW

File: bytecomp/translcore.ml [hunk 1]
```diff
@@ -438,31 +438,33 @@ let transl_primitive loc p env ty path =
   match prim with
   | Plazyforce ->
       let parm = Ident.create "prim" in
-      Lfunction{kind = Curried; params = [parm];
+      Lfunction{kind = Curried; params = [parm, Pgenval]; return = Pgenval;
                 body = Matching.inline_lazy_force (Lvar parm) Location.none;
                 loc = loc;
-                attr = default_stub_attribute }
+                attr = default_function_attribute }
```
Comment (mshinwell) [inline bytecomp/translcore.ml]: Is this an unrelated change?  (Same on line 452 and in other places below.)

File: bytecomp/simplif.ml [hunk 2]
```diff
@@ -457,13 +457,13 @@ let simplify_lets lam =
...
   | Lapply ap -> Lapply {ap with ap_func = simplif ap.ap_func;
                                  ap_args = List.map simplif ap.ap_args}
-  | Lfunction{kind; params; body = l; attr; loc} ->
+  | Lfunction{kind; params; return; body = l; attr; loc} ->
       begin match simplif l with
-        Lfunction{kind=Curried; params=params'; body; attr; loc}
+        Lfunction{kind=Curried; params=params'; return; body; attr; loc}
         when kind = Curried && optimize ->
-          Lfunction{kind; params = params @ params'; body; attr; loc}
+          Lfunction{kind; params = params @ params'; return; body; attr; loc}
```
Comment (mshinwell) [inline bytecomp/simplif.ml]: It might just be worth a comment as to the fact that the [return] type from the second function is being used (correctly), rather than the shadowed [return].

File: bytecomp/translcore.ml [hunk 3]
```diff
@@ -1160,15 +1167,21 @@ and transl_apply ?(should_be_tailcall=false) ?(inlined = Default_inline)
         let body =
           match build_apply handle ((Lvar id_arg, optional)::args') l with
             Lfunction{kind = Curried; params = ids; body = lam; attr; loc} ->
```
Comment (mshinwell) [inline bytecomp/translcore.ml]: The workings of this whole function are rather obscure.  Are some of the `Pgenval` annotations here over-specific?  For example maybe it's the case that `return` should be bound on line 1169 and then used on line 1172?  (Even if it turns out to always be `Pgenval` this may be clearer.)  I think it might be worth adding some comments.

File: bytecomp/translcore.ml [hunk 4]
```diff
@@ -1202,16 +1218,22 @@ and transl_function loc untuplify_fn repr partial param cases =
               (Matching.flatten_pattern size c_lhs, c_guard, c_rhs))
             cases in
         let params = List.map (fun _ -> Ident.create "param") pl in
-        ((Tupled, params),
+        let tparams = List.map (fun p -> p, Pgenval) params in
```
Comment (mshinwell) [inline bytecomp/translcore.ml]: Don't we know something about the kind of the patterns here?

**Review by mshinwell — CHANGES_REQUESTED** [review]
I'm not completely certain about some of the changes in Translcore (from line 1100 onwards) as I've mentioned in the comments, but the rest looked fine to me modulo a couple of minor things.

File: bytecomp/translcore.ml (hunk 1 above)
Comment (chambart) [reply bytecomp/translcore.ml]: Oups. good catch.

This patch was extracted from a sequence started before the merge of the stub patch.
I will fix this

File: bytecomp/translcore.ml (hunk 3 above)
Comment (chambart) [reply bytecomp/translcore.ml]: I made the propagation explicit. I tried propagating enough information for this not to be always Pgenval, but this seems too complex for the benefit.

Comment (chambart) [reply bytecomp/translcore.ml]: Also I am not knowledgeable enough to be confident commenting this function.

File: bytecomp/translcore.ml (hunk 4 above)
Comment (chambart) [reply bytecomp/translcore.ml]: We do, and the following case was wrong also.

**Comment by bobzhang** [comment]
@chambart  this looks cool, would you also propogate that which argument is an optional?
I am working on a pass in bucklescript to unbox optional argument, with this information directly
from the compiler, the optimization could be running faster(faster compilation matters)
(people like optional argument, it is a nice feature to js people, but it comes with a non-trivial cost)

**Comment by chambart** [comment]
@bobzhang I plan to add other types to the value_kind type: for instance I have this patch around for array type propagation. https://github.com/chambart/ocaml-1/commit/c388c4d3da9d451c2499e932137c0470c06832da
You should create a PR to add the types you need in there.
